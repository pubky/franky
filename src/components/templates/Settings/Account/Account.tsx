'use client';

import { useState } from 'react';
import { useRouter } from 'next/navigation';
import * as Molecules from '@/molecules';
import * as Organisms from '@/organisms';
import * as Libs from '@/libs';
import { AUTH_ROUTES } from '@/app';

export interface AccountProps {
  className?: string;
}

export function Account({ className }: AccountProps) {
  const router = useRouter();
  const [loadingDownload, setLoadingDownload] = useState(false);
  const [progressDownload, setProgressDownload] = useState(0);
  const [disposableAccount] = useState(false);
  const [showDeleteDialog, setShowDeleteDialog] = useState(false);

  const handleSignOut = () => {
    router.push(AUTH_ROUTES.LOGOUT);
  };

  const handleDownloadData = async () => {
    setLoadingDownload(true);
    setProgressDownload(0);
    // TODO: Implement download data logic
    setTimeout(() => {
      setLoadingDownload(false);
      setProgressDownload(100);
    }, 2000);
  };

  const handleOpenDeleteDialog = () => {
    setShowDeleteDialog(true);
  };

  return (
    <>
      <Molecules.SettingsSectionCard className={className}>
        <Molecules.SettingsSection
          icon={Libs.LogOut}
          title="Sign out from Pubky"
          description="Sign out to protect your account from unauthorized access."
          buttonText="Sign out"
          buttonIcon={Libs.LogOut}
          buttonId="settings-sign-out-btn"
          buttonOnClick={handleSignOut}
        />

        <Molecules.SettingsDivider />

        <Molecules.SettingsSection
          icon={Libs.Pencil}
          title="Edit your profile"
          description="Update your bio or user picture, so friends can find you easier."
          buttonText="Edit profile"
          buttonIcon={Libs.Pencil}
          buttonId="edit-profile-btn"
          buttonOnClick={() => {}}
        />

        <Molecules.SettingsDivider />

        <Molecules.SettingsSection
          icon={Libs.Lock}
          title="Back up your account"
          description={
            disposableAccount
              ? 'Without a backup you lose your account if you close your browser!'
              : 'You have already completed the backup, or closed your browser before doing so. Your recovery file and seed phrase have been deleted.'
          }
          buttonText="Back up account"
          buttonIcon={Libs.Lock}
          buttonId="backup-account-btn"
          buttonDisabled={!disposableAccount}
          buttonOnClick={() => {}}
        />

        <Molecules.SettingsDivider />

        <Molecules.SettingsSection
          icon={Libs.Download}
          title="Download your data"
          description="Your data on Pubky is yours. Export your account data to use it elsewhere."
          buttonText={loadingDownload ? `Downloading... ${progressDownload}%` : 'Download data'}
          buttonIcon={Libs.Download}
          buttonId="download-data-btn"
          buttonDisabled={loadingDownload}
          buttonOnClick={handleDownloadData}
        />

        <Molecules.SettingsDivider />

        <Molecules.SettingsSection
          icon={Libs.Trash2}
          title="Delete your account"
          description="Deleting your account will remove all of your posts, tags, profile information, contacts, custom streams, and settings or preferences."
          buttonText="Delete Account"
          buttonIcon={Libs.Trash2}
          buttonId="delete-account-btn"
          buttonVariant="destructive"
          titleClassName="text-destructive"
          iconClassName="text-destructive"
          buttonOnClick={handleOpenDeleteDialog}
        />
      </Molecules.SettingsSectionCard>

      <Organisms.DialogDeleteAccount isOpen={showDeleteDialog} onOpenChangeAction={setShowDeleteDialog} />
    </>
  );
}
