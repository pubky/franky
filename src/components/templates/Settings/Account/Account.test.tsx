import React from 'react';
import { describe, it, expect, vi } from 'vitest';
import { render, screen } from '@testing-library/react';
import { Account } from './Account';

// Mock next/navigation
vi.mock('next/navigation', () => ({
  useRouter: () => ({
    push: vi.fn(),
  }),
}));

describe('Account', () => {
  it('renders account content', () => {
    render(<Account />);
    expect(screen.getByText('Sign out from Pubky')).toBeInTheDocument();
  });

  it('renders all account sections', () => {
    render(<Account />);
    expect(screen.getByText('Sign out from Pubky')).toBeInTheDocument();
    expect(screen.getByText('Edit your profile')).toBeInTheDocument();
    expect(screen.getByText('Back up your account')).toBeInTheDocument();
    expect(screen.getByText('Download your data')).toBeInTheDocument();
    expect(screen.getByText('Delete your account')).toBeInTheDocument();
  });

  it('applies custom className', () => {
    const { container } = render(<Account className="custom-account" />);
    expect(container.firstChild).toBeInTheDocument();
  });
});

describe('Account - Snapshots', () => {
  it('matches snapshot', () => {
    const { container } = render(<Account />);
    expect(container.firstChild).toMatchSnapshot();
  });
});
