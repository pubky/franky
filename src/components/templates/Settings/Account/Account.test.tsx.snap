// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Account - Snapshots > matches snapshot 1`] = `
<div
  class="p-8 md:p-12 bg-[#1D1D20] rounded-md flex-col justify-start items-start gap-12 inline-flex"
>
  <div
    class="w-full flex-col justify-start items-start gap-3 flex"
  >
    <div
      class="flex-col justify-start items-start flex"
    >
      <div
        class="justify-start items-center gap-2 inline-flex pb-6"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-log-out"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m16 17 5-5-5-5"
          />
          <path
            d="M21 12H9"
          />
          <path
            d="M9 21H5a2 2 0 0 1-2-2V5a2 2 0 0 1 2-2h4"
          />
        </svg>
        <h2
          class="text-2xl font-bold text-foreground"
          data-testid="heading-2"
        >
          Sign out from Pubky
        </h2>
      </div>
      <p
        class="text-base font-medium leading-6 pb-6 text-[#D4D4DB]"
        data-testid="typography"
      >
        Sign out to protect your account from unauthorized access.
      </p>
      <button
        class="inline-flex items-center justify-center whitespace-nowrap text-sm transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*=\\"size-\\"])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/40 aria-invalid:border-destructive font-semibold cursor-pointer rounded-full border shadow-xs bg-secondary text-secondary-foreground hover:bg-secondary/80 h-10 gap-1 px-4 py-2 has-[>svg]:px-4"
        data-size="default"
        data-slot="button"
        data-variant="secondary"
        id="settings-sign-out-btn"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-log-out"
          fill="none"
          height="16"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m16 17 5-5-5-5"
          />
          <path
            d="M21 12H9"
          />
          <path
            d="M9 21H5a2 2 0 0 1-2-2V5a2 2 0 0 1 2-2h4"
          />
        </svg>
        Sign out
      </button>
    </div>
    <div
      aria-hidden="true"
      class="w-full h-px bg-white/10 my-6"
    />
    <div
      class="flex-col justify-start items-start flex"
    >
      <div
        class="justify-start items-center gap-2 inline-flex pb-6"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-pencil"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M21.174 6.812a1 1 0 0 0-3.986-3.987L3.842 16.174a2 2 0 0 0-.5.83l-1.321 4.352a.5.5 0 0 0 .623.622l4.353-1.32a2 2 0 0 0 .83-.497z"
          />
          <path
            d="m15 5 4 4"
          />
        </svg>
        <h2
          class="text-2xl font-bold text-foreground"
          data-testid="heading-2"
        >
          Edit your profile
        </h2>
      </div>
      <p
        class="text-base font-medium leading-6 pb-6 text-[#D4D4DB]"
        data-testid="typography"
      >
        Update your bio or user picture, so friends can find you easier.
      </p>
      <button
        class="inline-flex items-center justify-center whitespace-nowrap text-sm transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*=\\"size-\\"])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/40 aria-invalid:border-destructive font-semibold cursor-pointer rounded-full border shadow-xs bg-secondary text-secondary-foreground hover:bg-secondary/80 h-10 gap-1 px-4 py-2 has-[>svg]:px-4"
        data-size="default"
        data-slot="button"
        data-variant="secondary"
        id="edit-profile-btn"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-pencil"
          fill="none"
          height="16"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M21.174 6.812a1 1 0 0 0-3.986-3.987L3.842 16.174a2 2 0 0 0-.5.83l-1.321 4.352a.5.5 0 0 0 .623.622l4.353-1.32a2 2 0 0 0 .83-.497z"
          />
          <path
            d="m15 5 4 4"
          />
        </svg>
        Edit profile
      </button>
    </div>
    <div
      aria-hidden="true"
      class="w-full h-px bg-white/10 my-6"
    />
    <div
      class="flex-col justify-start items-start flex"
    >
      <div
        class="justify-start items-center gap-2 inline-flex pb-6"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-lock"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <rect
            height="11"
            rx="2"
            ry="2"
            width="18"
            x="3"
            y="11"
          />
          <path
            d="M7 11V7a5 5 0 0 1 10 0v4"
          />
        </svg>
        <h2
          class="text-2xl font-bold text-foreground"
          data-testid="heading-2"
        >
          Back up your account
        </h2>
      </div>
      <p
        class="text-base font-medium leading-6 pb-6 text-[#D4D4DB]"
        data-testid="typography"
      >
        You have already completed the backup, or closed your browser before doing so. Your recovery file and seed phrase have been deleted.
      </p>
      <button
        class="inline-flex items-center justify-center whitespace-nowrap text-sm transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*=\\"size-\\"])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/40 aria-invalid:border-destructive font-semibold cursor-pointer rounded-full border shadow-xs bg-secondary text-secondary-foreground hover:bg-secondary/80 h-10 gap-1 px-4 py-2 has-[>svg]:px-4"
        data-size="default"
        data-slot="button"
        data-variant="secondary"
        disabled=""
        id="backup-account-btn"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-lock"
          fill="none"
          height="16"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <rect
            height="11"
            rx="2"
            ry="2"
            width="18"
            x="3"
            y="11"
          />
          <path
            d="M7 11V7a5 5 0 0 1 10 0v4"
          />
        </svg>
        Back up account
      </button>
    </div>
    <div
      aria-hidden="true"
      class="w-full h-px bg-white/10 my-6"
    />
    <div
      class="flex-col justify-start items-start flex"
    >
      <div
        class="justify-start items-center gap-2 inline-flex pb-6"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-download"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12 15V3"
          />
          <path
            d="M21 15v4a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2v-4"
          />
          <path
            d="m7 10 5 5 5-5"
          />
        </svg>
        <h2
          class="text-2xl font-bold text-foreground"
          data-testid="heading-2"
        >
          Download your data
        </h2>
      </div>
      <p
        class="text-base font-medium leading-6 pb-6 text-[#D4D4DB]"
        data-testid="typography"
      >
        Your data on Pubky is yours. Export your account data to use it elsewhere.
      </p>
      <button
        class="inline-flex items-center justify-center whitespace-nowrap text-sm transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*=\\"size-\\"])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/40 aria-invalid:border-destructive font-semibold cursor-pointer rounded-full border shadow-xs bg-secondary text-secondary-foreground hover:bg-secondary/80 h-10 gap-1 px-4 py-2 has-[>svg]:px-4"
        data-size="default"
        data-slot="button"
        data-variant="secondary"
        id="download-data-btn"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-download"
          fill="none"
          height="16"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12 15V3"
          />
          <path
            d="M21 15v4a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2v-4"
          />
          <path
            d="m7 10 5 5 5-5"
          />
        </svg>
        Download data
      </button>
    </div>
    <div
      aria-hidden="true"
      class="w-full h-px bg-white/10 my-6"
    />
    <div
      class="flex-col justify-start items-start flex"
    >
      <div
        class="justify-start items-center gap-2 inline-flex pb-6"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-trash2 lucide-trash-2 text-destructive"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M3 6h18"
          />
          <path
            d="M19 6v14c0 1-1 2-2 2H7c-1 0-2-1-2-2V6"
          />
          <path
            d="M8 6V4c0-1 1-2 2-2h4c1 0 2 1 2 2v2"
          />
          <line
            x1="10"
            x2="10"
            y1="11"
            y2="17"
          />
          <line
            x1="14"
            x2="14"
            y1="11"
            y2="17"
          />
        </svg>
        <h2
          class="text-2xl font-bold text-destructive"
          data-testid="heading-2"
        >
          Delete your account
        </h2>
      </div>
      <p
        class="text-base font-medium leading-6 pb-6 text-[#D4D4DB]"
        data-testid="typography"
      >
        Deleting your account will remove all of your posts, tags, profile information, contacts, custom streams, and settings or preferences.
      </p>
      <button
        class="inline-flex items-center justify-center whitespace-nowrap text-sm transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*=\\"size-\\"])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-[3px] aria-invalid:ring-destructive/40 aria-invalid:border-destructive font-semibold cursor-pointer rounded-full border shadow-xs bg-destructive/60 text-destructive-foreground hover:bg-destructive/90 focus-visible:ring-destructive/40 h-10 gap-1 px-4 py-2 has-[>svg]:px-4"
        data-size="default"
        data-slot="button"
        data-variant="destructive"
        id="delete-account-btn"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-trash2 lucide-trash-2"
          fill="none"
          height="16"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M3 6h18"
          />
          <path
            d="M19 6v14c0 1-1 2-2 2H7c-1 0-2-1-2-2V6"
          />
          <path
            d="M8 6V4c0-1 1-2 2-2h4c1 0 2 1 2 2v2"
          />
          <line
            x1="10"
            x2="10"
            y1="11"
            y2="17"
          />
          <line
            x1="14"
            x2="14"
            y1="11"
            y2="17"
          />
        </svg>
        Delete Account
      </button>
    </div>
  </div>
</div>
`;
